### PR TITLE
[Tabular] Support different random seeds per fold

### PR DESCRIPTION
**Pull request by LennartPurucker**
This PR ensures that each fold gets a different random seed. I implemented it as part of the kwargs to avoid users setting random seeds as a hyperparameter. 

> By default, we keep a static seed of 0 to ensure reproducibility. When using a bagged model, a static seed of 0 to n_splits-1 is used to ensure that each fold has a different seed that is consistent across different models.

Note, one can disable this behavior and use a static seed across all folds (similar to the prior implementation via):

TODO:

- [x] Add code to support this for all other models. 

--- 
One can control the seed behavior via `ag_args_ensemble`:

```python
# Enable behavior of varying the seeds across folds
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10}},
    ag_args_ensemble={"fold_fitting_strategy": 'sequential_local', "vary_seed_across_folds": True}
)

# Set your own random seed for all models
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10}},
    ag_args_ensemble={"fold_fitting_strategy": 'sequential_local', "model_random_seed": 42}
)

# Set your own random seed per configuration
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10, "ag_args_ensemble": {"model_random_seed": 42}}},
)

```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 1]
```diff
@@ -60,6 +60,8 @@ class BaggedEnsembleModel(AbstractModel):
     """
 
     _oof_filename = "oof.pkl"
+    _vary_seed_across_folds: bool = True
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: can be an `ag_args_fit` param to `BaggedEnsembleModel` with default True.

this can then be specified via `ag_args_ensemble` by the user.

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py [hunk 2]
```diff
@@ -285,8 +284,16 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
...
+
+        if "seed_value" in params:
+            logger.log(
+                40,
+                "\t'seed_value' found in hyperparamters is ignored!"
+                "AutoGluon controls the random seed for the model training.",
+            )
+
+        if random_seed is not None:
+            train_params["params"]["seed"] = random_seed
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: I wonder if we should have a dedicated method for handling seed so it is easier to check if a model has implemented seed support.

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py [hunk 3]
```diff
@@ -285,8 +284,16 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
             train_params["params"]["num_classes"] = self.num_classes
         elif self.problem_type == QUANTILE:
             train_params["params"]["quantile_levels"] = self.quantile_levels
-        if seed_val is not None:
-            train_params["params"]["seed"] = seed_val
+
+        if "seed_value" in params:
+            logger.log(
+                40,
+                "\t'seed_value' found in hyperparamters is ignored!"
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: nit: hyperparameters

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py [hunk 4]
```diff
@@ -285,8 +284,16 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
...
+                "AutoGluon controls the random seed for the model training.",
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: Probably not ideal. Users should be able to specify random seeds. If they specify it, then we just use it instead of `random_seed`.

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 5]
```diff
@@ -1034,6 +1034,11 @@ def fit(self, *, log_resources: bool = False, **kwargs):
             verbosity 2: logs only important information.
             verbosity 1: logs only warnings and exceptions.
             verbosity 0: logs only exceptions.
+        random_seed : int, default = 0
+            The random seed used to control the randomness during fitting the model.
+            By default, we keep a static seed of 0 to ensure reproducibility.
+            When using a bagged model, a static seed of 0 to n_splits-1 is used to ensure that each fold has a
+            different seed that is consistent across different models.
```
Comment (Innixma) [inline core/src/autogluon/core/models/abstract/abstract_model.py]: I'm a bit torn here. random_seed is a hyperparameter, but is not being treated as one by this logic. Instead we could consider using `ag_args_fit` to store the value?

Comment (LennartPurucker) [reply core/src/autogluon/core/models/abstract/abstract_model.py]: Mhm, also to the above, I think we need to answer for ourselves if we want to treat random_seed as an HP.

In all cases where this can be helpful, it indicates that the model is extremely unstable during training. Not sure if we will ever have such a model. At the same time, one could give each HP config a different (start) seed and thus likely get better (val-overfitted) MLPs. 

Adding the random_seed start valuet to ag_args_fit

**Comment by LennartPurucker** [comment]
@Innixma I have reworked the PR based on your feedback. Again, with some larger design choices that need your feedback. 

Moreover, I have added a test case and am now extending it to all other models in AG

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 6]
```diff
@@ -108,6 +108,8 @@ def _set_default_params(self):
             "stratify": "auto",
             "bin": "auto",
             "n_bins": None,
+            "vary_seed_across_folds": True, # If True, the seed used for each fold will be varied across folds.
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: For this PR, lets set it to default False so we enable the ability to use the logic, but make it so the logic doesn't yet change AutoGluon's default behavior until we benchmark things and decide if it is beneficial.

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 7]
```diff
@@ -938,9 +945,13 @@ def _generate_fold_configs(
                     folds_to_fit=folds_to_fit,
                     folds_finished=fold - fold_start,
                     folds_left=fold_end - fold,
+                    random_seed=random_seed_value
                 )
 
                 fold_fit_args_list.append(fold_ctx)
+                if vary_seed_across_folds:
+                    random_seed_value += 1
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: This is technically incorrect in the scenario where `n_repeat_start` and `fold_in_set_start` don't equal 0, such as when fitting an extra repeat of the bag in a future fit call.

The way to do it correctly would be to ensure the Nth child model gets a random_seed_value of `n-1`. Code will be a bit more complicated, but pretty sure there is a smart way to do it, as I've already had to deal with this same problem in the past somewhere in the bagged ensemble code.

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py [hunk 8]
```diff
@@ -128,7 +128,12 @@ def _estimate_memory_usage_static(
         approx_mem_size_req = data_mem_usage_bytes + histogram_mem_usage_bytes + mem_size_estimators
         return approx_mem_size_req
 
-    def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cpus=0, sample_weight=None, sample_weight_val=None, verbosity=2, **kwargs):
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict | None = None) -> tuple[bool, int | None]:
+        if "seed_value" in hyperparameters:
+            return True, hyperparameters["seed_value"]
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: Is it only `seed_value`? Unsure why, but the AG code also uses `seed` as in:

```python
        if seed_val is not None:
            train_params["params"]["seed"] = seed_val
```

Maybe I wasn't comprehensive when I originally coded it.

File: tabular/tests/unittests/models/advanced/test_model_random_seed.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    Tests 2 fit types:
+    sequential fit + sequential bag
+    sequential fit + parallel bag
+    parallel fit + parallel bag
+    parallel fit + sequential bag
+    """
+    sample_size = 50
+    dataset_name = "adult"
+    directory_prefix = "./datasets/"
+    train_data, _, dataset_info = FitHelper.load_dataset(name=dataset_name, directory_prefix=directory_prefix)
```
Comment (Innixma) [inline tabular/tests/unittests/models/advanced/test_model_random_seed.py]: nit: we can use the toy datasets I generate for the other unit tests to be faster and avoid an s3 download.

```
dataset_name = "toy_binary"
train_data, test_data, dataset_info = FitHelper.load_dataset(name=dataset_name, directory_prefix=directory_prefix)
```

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py [hunk 10]
```diff
@@ -285,8 +289,10 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
             train_params["params"]["num_classes"] = self.num_classes
         elif self.problem_type == QUANTILE:
             train_params["params"]["quantile_levels"] = self.quantile_levels
-        if seed_val is not None:
-            train_params["params"]["seed"] = seed_val
+
+        self.init_random_seed(random_seed=random_seed, hyperparameters=params)
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: I'd like to avoid having the model developer always need to call this. Why not do it in `fit` instead of `_fit`? That way we can do it alongside all the other initialization logic. Maybe in the `self._init_params()` call?

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 11]
```diff
@@ -800,6 +802,8 @@ def _fit_folds(
             k_fold_end=k_fold_end,
             n_repeat_start=n_repeat_start,
             n_repeat_end=n_repeats,
+            vary_seed_across_folds=self.params["vary_seed_across_folds"],
+            random_seed_start_value=self.params["model_random_seed"],
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: So currently BaggedEnsembleModel ignores `random_seed` in its own `_fit` call. This seems strange, as that means that nested `BaggedEnsembleModel` would not handle `random_seed` properly, as whatever is specified by the parent `BaggedEnsembleModel` would be ignored by the children `BaggedEnsembleModel`.

Maybe we don't care enough to address this gap, but good to document it if we decide not to fix this.

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py (hunk 10 above)
Comment (LennartPurucker) [reply tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: ~~Okay, but do we have the correct hyperparameters at that time? 
I was not sure but thought that some model might only get them in the `_fit` call.~~

Should work

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py (hunk 8 above)
Comment (LennartPurucker) [reply tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: Fixing both is needed for correct backward compatibility.
Here, seed_val was overwritting seed

File: tabular/src/autogluon/tabular/models/lgb/lgb_model.py (hunk 10 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/lgb/lgb_model.py]: correct hyperparameter values are determined in the `self._init_params()` call, as that is when they are set

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 12]
```diff
@@ -109,12 +109,15 @@ def _set_default_params(self):
...
             "vary_seed_across_folds": False, # If True, the seed used for each fold will be varied across folds.
-            "model_random_seed": 0,  # Start value for the random seed used for the seeds of the fold models.
+            "model_random_seed": 0,
         }
         for param, val in default_params.items():
             self._set_default_param_value(param, val)
         super()._set_default_params()
 
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict | None = None) -> int | None | str:
+        return hyperparameters.get("model_random_seed", "N/A")
```
Comment (Innixma) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: Crashes if `hyperparameters is None`?

Comment (LennartPurucker) [reply core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: Ah nvm, should only be called if it is not None, fixing types...

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py (hunk 11 above)
Comment (LennartPurucker) [reply core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: This case should be supported now as well

**Comment by LennartPurucker** [comment]
Important to note: when we refit, we reset the seed to the default value, as we are not in a bagged mode anymore. That should be fine however, as we do not know the "correct" seed to set for refitting. Moreover, models with `use_child_oof=True` will also only use the default seed.

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py (hunk 7 above)
Comment (LennartPurucker) [reply core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py]: Should be fixed with the latest update. If I understand the code correctly, `fold` gives the `n` of all splits.

**Comment by LennartPurucker** [comment]
The current version should now be close to the final version

**Comment by Innixma** [comment]
looks like test is failing

**Comment by LennartPurucker** [comment]
Should be fixed now

**Comment by LennartPurucker** [comment]
The latest version now also supports child_oof and refit_folds to have different seeds if one passed `model_random_seed` to `ag_args_ensemble`.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 13]
```diff
@@ -164,6 +164,10 @@ def _prepare_params(params):
 
         return processor_kwargs, optimizer_kwargs, fit_kwargs, loss_kwargs, params
 
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict) -> int | None | str:
+        return hyperparameters.get("seed_value", "N/A")
+
+
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py]:
```suggestion
```
